feat: additional rule metadata for deprecations

DMartens opened the pull request:
## Summary
This is a continuation of #116 by @bmish to specify additional properties for rule deprecations.
The main changes are:
- move all properties under `meta.deprecated` and deprecate `meta.replacedBy`
- "shorthands" for properties
- additional properties: `kind`, `deprecatedSince`, `availableUntil` 
- more open questions

## Related Issues
> * [Change Request: Support additional rule metadata for deprecations eslint#18061](https://github.com/eslint/eslint/issues/18061)

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+  url: string // URL to more information about this deprecation in general.
+}
+
+type ReplacementKind =
+  'moved' |  // The rule has moved to another plugin if plugin is set, otherwise the rule is renamed in the same plugin
+  'merged' | // The rule merged with another rule
+  'option'   // The current rule behavior is available as an option in the replacement rule
```
I think `ReplacedByInfo.kind` isn't necessary as this info and more context can be provided as part of the `ReplacedByInfo.info` (or ReplacedByInfo.info.message) text.

DMartens replied:
The idea behind this property was:
- for documentation generators: able to provide a default "message" without having to specify the same message for multiple rules (e.g. the rule was renamed from "foo" to "bar")
- a possible config migration tool: suggest changes without relying on a specific format in info.message (e.g. rename the rule in the config)

nzakas replied:
I'm also not sure of the value here. A single word just doesn't given enough context. I also fear this would become a maintenance issue as we'll end finding other reasons rules were deprecated and need to update this list with more unclear terms.

mdjermanovic replied:
Looks like we're in favor of removing this property. @DMartens can you update the RFC?

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+  rule?: string | Specifier
+  info?: string | Message
+  kind?: ReplacementKind // Defaults to "moved" if missing
+  deprecatedSince?: Version // Helps users gauge when to migrate and useful for documentation
+  availableUntil?: Version | null // The estimated version when the rule is removed (probably the next major version). null means the rule is "frozen" (will be available but will not be changed)
```
Shouldn't these two properties rather be on `DeprecateInfo`, as they don't seem related to particular replacements but the deprecation itself?

DMartens replied:
Yes, they should be.

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
````diff
@@ -0,0 +1,281 @@
+
+We can also support the same `meta.deprecated` and `meta.replacedBy` properties on configurations and processors (the other kinds of objects exported by ESLint plugins), replacing `rule` with `config` or `processor` as needed. This would be part of the effort to standardize documentation properties in <https://github.com/eslint/eslint/issues/17842>.
+
+### Shorthand
+The shorthand for the properties `plugin`, `rule` and `info` is just a string representing either the `name`/`message` or the `url` based on its content.
+If it starts with a protocol (e.g. `https://`) the property should be interpreted as if only the `url` property is set, otherwise it should be interpreted as `name`/`message` property.
+This shorthand also applies for the existing `meta.deprecated` which then applies for the `meta.deprecated.info` properties.
+Some examples:
+```js
+{ meta: { deprecated: { plugin: 'https://eslint.style' } } }   // <=> { meta: { deprecated: { plugin: { url: 'https://eslint.style' } } } }
+{ meta: { deprecated: { plugin: '@eslint-stylistic/js' } } }   // <=> { meta: { deprecated: { plugin: { name: '@eslint-stylistic/js' } } } }
+{ meta: { deprecated: 'https://eslint.style/guide/migration' } // <=> { meta: { deprecated: { info: { url: 'https://eslint.style/guide/migration' } } } }
+```
````
In my opinion, this is an unnecessary complexity that tooling would need to support. I'd be in favor of dropping shorthands to simplify this. Deprecations don't happen often, and it doesn't take much effort to write an object with a `name` or `url` property instead of a string.

nzakas replied:
Agree with @mdjermanovic on this.

mdjermanovic replied:
`deprecated: {}` can be interpreted the same as `deprecated: true`. Other properties can be interpreted as if they were not provided.

mdjermanovic replied:
Looks like we all agree to drop the shorthands. @DMartens can you update the RFC?

DMartens replied:
Dropped the shorthands and made the object properties optional.

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+
+1. Is there additional deprecation information we'd like to represent? Note that additional information can always be added later, but it's good to consider any possible needs now.
+2. Should `meta.deprecated.plugin.id` accommodate different package registries (e.g. [jsr](https://jsr.io/) with `jsr:eslint-plugin-example`)
+3. Should the exact regular expression for the shorthand which decides whether it is a description or URL be specified?
+4. Should the shorthand also be applied for the string form of the `meta.deprecated` property?
```
I'd prefer dropping shorthands, so we wouldn't have these questions.

DMartens commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+2. Should `meta.deprecated.plugin.id` accommodate different package registries (e.g. [jsr](https://jsr.io/) with `jsr:eslint-plugin-example`)
+3. Should the exact regular expression for the shorthand which decides whether it is a description or URL be specified?
+4. Should the shorthand also be applied for the string form of the `meta.deprecated` property?
+5. Which "extension points" (rules, processors, configurations, parsers, formatters) shold be supported?
+6. Should the `rule` key be dependent on the "extension point" (e.g. `processor` for processors) or renamed (e.g. ``) so that it is the same property name for all?
```
I do not know of examples for other extension points being deprecated but especially with the introduction of language plugins, I can see some processors being deprecated (e.g. eslint-plugin-markdown).

nzakas commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+};
+
+/* At least one property is required */
+type DeprecateInfo = {
```
```suggestion
type DeprecatedInfo = {
```

nzakas commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+  kind?: ReplacementKind // Defaults to "moved" if missing
+}
+
+type Message = {
```
I'm not sure this type is necessary. I'd rather just move `message` and `url` to `DeprecatedInfo` and `ReplacedByInfo`. Remove one level of objects.

mdjermanovic replied:
I also think it would be better to simplify this by moving these properties directly into `DeprecatedInfo` and `ReplacedByInfo`.

nzakas commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+          name: '@stylistic/js',
+          url: 'https://eslint.style/',
+        },
+        rule: 'https://eslint.style/rules/js/semi',
+        },
```
Syntax error

nzakas commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+module.exports = {
+  meta: {
+    deprecated: {
+      message: 'Stylistic rules are being moved out of ESLint core.',
+      url: 'https://eslint.org/blog/2023/10/deprecating-formatting-rules/',
```
`message` and `url` aren't listed as members above (although, I do like this better. See my earlier comments.)

nzakas commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+-->
+
+1. Is there additional deprecation information we'd like to represent? Note that additional information can always be added later, but it's good to consider any possible needs now.
+2. Should `meta.deprecated.plugin.id` accommodate different package registries (e.g. [jsr](https://jsr.io/) with `jsr:eslint-plugin-example`)
```
```suggestion
2. Should `meta.deprecated.plugin.name` accommodate different package registries (e.g. [jsr](https://jsr.io/) with `jsr:eslint-plugin-example`)
```

nzakas replied:
In general, I don't think we should worry about package registries. People should include a URL to the package wherever its published if this is a concern.

nzakas commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,281 @@
+type DeprecateInfo = {
+  info?: Message
+  replacedBy?: (string|ReplacedByInfo)[] // An empty array explicitly states that there is no replacement
+  deprecatedSince?: Version // Helps users gauge when to migrate and useful for documentation
+  availableUntil?: Version | null // The estimated version when the rule is removed (probably the next major version). null means the rule is "frozen" (will be available but will not be changed)
```
> Also, let's suppose that we decide to deprecate a rule in ESLint v10. What value should be set for `deprecatedSince`? Should it be `"v10.0.0-alpha.0"`,`"v10.0.0"`, `"v10"`, etc.?

`10.0.0`. Prereleases don't count and we should use the full semver version.

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,267 @@
+
+type Message = {
+  message?: string // General message presented to the user. Content depends on the property (e.g. for the key rule why the rule is deprecated or for info how to replace the rule)
+  url?:            // URL to more information about this deprecation in general.
```
```suggestion
  url?: string           // URL to more information about this deprecation in general.
```

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,267 @@
+}
+
+type Specifier = {
+  name?: string // Name of the rule / configuration / ...
+  url?: string // URL to more information about this deprecation in general.
```
```suggestion
  name?: string // Name of the plugin / rule / ...
  url?: string  // URL to documentation for the plugin / rule / ...
```

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,267 @@
+- Update ESLint's website generator to take into account the additional information for rule doc deprecation notices
+- Update [LintResult.usedDeprecatedRules](https://github.com/eslint/eslint/blob/0f5df509a4bc00cff2c62b90fab184bdf0231322/lib/eslint/eslint.js#L197-L211)
+
+External changes:
+
+- Update the [types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b77d83e019025017b06953907cb77f35e4231714/types/eslint/index.d.ts#L734) in @types/eslint
```
This would be a change inside ESLint, because the `eslint` package now includes types: 

https://github.com/eslint/eslint/blob/7dd402d33226d821a17b22c4753ce9c40fc909bd/lib/types/index.d.ts#L744-L747

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,267 @@
+  'merged' | // The rule merged with another rule
+  'option'   // The current rule behavior is available as an option in the replacement rule
+
+/* Version string of the package containing the rule */
```
```suggestion
/* Version string. It should be the full semver version without the "v" prefix, e.g. "8.53.0" */
```

Per the discussion on https://github.com/eslint/rfcs/pull/124#discussion_r1773117929

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,267 @@
+- Ensure these properties are allowed on configurations, parsers and processors
+- Add any additional information to these properties in core rules as desired (such as in <https://github.com/eslint/eslint/issues/18053>)
+- Update ESLint's website generator to take into account the additional information for rule doc deprecation notices
+- Update [LintResult.usedDeprecatedRules](https://github.com/eslint/eslint/blob/0f5df509a4bc00cff2c62b90fab184bdf0231322/lib/eslint/eslint.js#L197-L211)
```
Can you add more details about this change? Shall we keep the existing `replacedBy: string[]` property and/or add a new one with the `DeprecatedInfo` object?

fasttime replied:
I think that removing or changing the format of `DeprecatedRuleInfo.replacedBy` could break existing tools like custom formatters. Maybe it would be better to expose the `RuleMeta.deprecated` object as a new property of `DeprecatedRuleInfo`?

DMartens replied:
I think we should normalize the old and new format for `replacedBy` and provide the new data if available under a new property named `info` as this backwards-compatible.

mdjermanovic replied:
> I think we should normalize the old and new format for `replacedBy` and provide the new data if available under a new property named `info` as this backwards-compatible.

Sounds good to me.

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,267 @@
+In terms of actual changes inside ESLint needed for this:
+
+- Mention the new schema in the [custom rule documentation](https://eslint.org/docs/latest/extend/custom-rules#rule-structure)
+- Ensure these properties are allowed on configurations, parsers and processors
```
If this RFC only aims to add support for additional metadata for rule deprecations, then I think this change is out of its scope.

DMartens replied:
I focused on the rule case as the original proposal did.
As mentioned in FAQ question 3, I think that the remaining extension points (e.g. configs) also should have a deprecated property.
In the case of configs, ESLint could print an error informing the user of the deprecated shared config and what the user should do.
I would add additional sections for them and extend the current ones (e.g. internal changes) if other extension points should be part of the proposal.

mdjermanovic replied:
@eslint/eslint-tsc do we want this RFC to cover configurations, parsers, and processors as well, or only rules?

mdjermanovic replied:
Then, I think we should remove this line from the RFC as it's out of scope and it isn't clear what the action item would be.

fasttime commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,267 @@
+      replacedBy: [
+        {
+          plugin: {
+          name: '@stylistic/js',
+          url: 'https://eslint.style/',
+        },
+        rule: 'https://eslint.style/rules/js/semi',
```
```suggestion
            name: '@stylistic/js',
            url: 'https://eslint.style/',
          },
          rule: 'https://eslint.style/rules/js/semi',
```

To fix the indentation.

fasttime commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,267 @@
+    you can remove this section.
+-->
+
+1. Is there additional deprecation information we'd like to represent? Note that additional information can always be added later, but it's good to consider any possible needs now.
```
> Is there additional deprecation information we'd like to represent?

Not for ESLint directly as far as I know, but third-party plugins may want to include additional deprecation information for their specific purposes. One use case that's been already mentioned in https://github.com/eslint/rfcs/pull/124/files#r1764772326 is the ability to mark a rule that was renamed. Another use case is to include the commit hash or PR link in which a rule was deprecated, like shown in the `typescript-eslint` docs ([an example](https://typescript-eslint.io/rules/block-spacing/)).

My suggestion is that we only focus on information consumed by ESLint in this RFC (hence no `ReplacementKind` needed). But I'd also suggest that we don't try to validate or forbid unrecognized properties in a `meta.deprecated` object.

DMartens commented:
Sorry for dropping the ball on this.
I have incorporated the following suggestions:
  - Rename introduced main type
  - Removed replacement kind
  - Make info property (message, url) root properties in DeprecatedInfo and ReplacedInfo
  - Remove syntax error in example
  - Add answer for FAQ question 2
  - Add note about version format
  - Reference internal ESLint types instead of DefinitelyTyped
  - Mention plugin developers can extend the types
  - Detailed the changes to usedDeprecatedRules

I am still committed to finishing this proposal.

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,258 @@
+- Ensure these properties are allowed on all extension points (configurations, formatter, parsers and processors)
+- Add any additional information to these properties in core rules as desired (such as in <https://github.com/eslint/eslint/issues/18053>)
+- Update ESLint's website generator to take into account the additional information for rule doc deprecation notices
+- Update [LintResult.usedDeprecatedRules](https://github.com/eslint/eslint/blob/0f5df509a4bc00cff2c62b90fab184bdf0231322/lib/eslint/eslint.js#L197-L211) by normalizing the old and new format for the existing `replacedBy` property and adding a new property with the name `info` for rules using the new deprecated format
+
+External changes:
+
+- Update the [types](https://github.com/eslint/eslint/blob/35a8858d62cb050fa0b56702e55c94ffaaf6956d/lib/types/index.d.ts#L745) in eslint
```
```suggestion
- Update [LintResult.usedDeprecatedRules](https://github.com/eslint/eslint/blob/0f5df509a4bc00cff2c62b90fab184bdf0231322/lib/eslint/eslint.js#L197-L211) by normalizing the old and new format for the existing `replacedBy` property and adding a new property with the name `info` for rules using the new deprecated format
- Update the [types](https://github.com/eslint/eslint/blob/35a8858d62cb050fa0b56702e55c94ffaaf6956d/lib/types/index.d.ts#L745) in eslint

External changes:

```

Since this in an internal change.

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,258 @@
+      url: 'https://eslint.org/blog/2023/10/deprecating-formatting-rules/',
+      replacedBy: [
+        {
+          plugin: {
+          name: '@stylistic/js',
+          url: 'https://eslint.style/rules/js/semi',
+        },
```
This should be a URL to the plugin docs, and there should be an additional `rule` key with rule name and link to the rule docs?

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,258 @@
+type DeprecatedInfo = {
+  message?: string // General message presented to the user, e.g. for the key rule why the rule is deprecated or for info how to replace the rule
+  url?: string     // URL to more information about this deprecation in general
+  replacedBy?: (string|ReplacedByInfo)[] // An empty array explicitly states that there is no replacement
```
Do we need `string|` here?

DMartens replied:
No, this was part of the "shorthands" which I seem to have missed.

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,262 @@
+      message: 'Stylistic rules are being moved out of ESLint core.',
+      url: 'https://eslint.org/blog/2023/10/deprecating-formatting-rules/',
+      replacedBy: [
+        {
+          plugin: {
+          name: '@stylistic/js',
+          url: 'https://eslint.style/',
+          rule: {
+            name: 'semi',
+            url: 'https://eslint.style/rules/js/semi'
+          }
+        },
```
```suggestion
        {
          plugin: {
            name: "@stylistic/eslint-plugin-js",
            url: "https://eslint.style/",
          },
          rule: {
            name: "semi",
            url: "https://eslint.style/rules/js/semi",
          },
        },
```

Per the types, `rule` is not inside `plugin`.

mdjermanovic commented on designs/2024-deprecated-rule-metadata/README.md:
```diff
@@ -0,0 +1,262 @@
+  - <https://github.com/eslint/eslint/issues/5774#issuecomment-220640368>
+- Related
+  - <https://github.com/eslint/eslint/issues/17842>
+  - <https://github.com/eslint/eslint/issues/18694>
```
```suggestion
```

I think this isn't related. Frozen rules are not deprecated.

DMartens replied:
I added this reference for `availableUntil: null` which should signify that a rule is frozen.

mdjermanovic replied:
The linked issue is only about marking some non-deprecated rules as frozen, so I still think it isn't related to the subject of this RFC.

frozen = we're not accepting new features (e.g., new options), but the rule is actively maintained (bug fixes & support for new syntax).

deprecated = we're not making any changes, and we don't guarantee that the rule won't be removed in a future major version:

https://eslint.org/docs/latest/use/rule-deprecation

mdjermanovic reviewed (commented):
Just a few small suggestions, then LGTM.

mdjermanovic reviewed (approved):
Looks good to me, thanks!

nzakas reviewed (approved):
LGTM. Moving to final commenting.

fasttime reviewed (approved):
LGTM.

bmish commented:
Thanks for finishing this up @DMartens!
